### PR TITLE
added Graphormer.jl

### LINKED ISSUE #251: implement Graphormer
https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf

### PR DESCRIPTION
**Pull request by 5hv5hvnk**
fixes #251 
To do:
- [x] struct based layers
- [x] doc-string
- [ ] example pipeline
- [ ] unit tests
- [ ] pass build tests

@CarloLucibello is there anything else to work on?

### CODE REVIEW

**Comment by 5hv5hvnk** [comment]
Is there a contributing guideline doc so I can fix the build tests at least?

File: src/Graphormer.jl [hunk 1]
```diff
@@ -0,0 +1,38 @@
+using Flux
+using Flux: @epochs, mse, params
+
+function Graphormer(input_dim, hidden_dim, output_dim, num_layers, num_heads)
```
Comment (CarloLucibello) [inline src/Graphormer.jl:4]: Please be consistent with the codebase. 
Layers should be structs, not functions, and should live in `src/layers/conv.jl`.

**Comment by CarloLucibello** [comment]
test failures is probably related to some changes in dependencies like FiniteDifferences.jl, I'll have to investigate

File: src/Graphormer.jl (hunk 1 above)
Comment (5hv5hvnk) [reply src/Graphormer.jl:4]: Sure let me update it accordingly

Comment (5hv5hvnk) [reply src/Graphormer.jl]: @CarloLucibello can you have a look at the updated script?

File: src/layers/conv.jl [hunk 2]
```diff
@@ -1670,3 +1670,67 @@ function Base.show(io::IO, l::TransformerConv)
     (in, ein), out = l.channels
     print(io, "TransformerConv(($in, $ein) => $out, heads=$(l.heads))")
 end
+
+struct GraphomerLayer{DX <: Dense, DE <: Union{Dense, Nothing}, T, A <: AbstractMatrix, F, B} <: GNNLayer
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1674]:
```suggestion
struct GraphormerLayer{DX <: Dense, DE <: Union{Dense, Nothing}, T, A <: AbstractMatrix, F, B} <: GNNLayer
```

File: src/layers/conv.jl [hunk 3]
```diff
@@ -1670,3 +1670,67 @@ function Base.show(io::IO, l::TransformerConv)
...
+
+    # a hand-written message passing
+    m = apply_edges((xi, xj, e) -> message(l, xi, xj, e), g, Wx, Wx, e)
+    α = softmax_edge_neighbors(g, m.logα)
+    β = α .* m.Wxj
+    x = aggregate_neighbors(g, +, β)
+
+    if !l.concat
+        x = mean(x, dims = 2)
+    end
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1736]: missing an `end` here

File: src/layers/conv.jl (hunk 2 above)
Comment (CarloLucibello) [inline src/layers/conv.jl:1674]: we can start adding a docstring and some tests

**Comment by CarloLucibello** [comment]
I think it would be good to add an example to the example folder showing the full pipeline, 
since there are some ingredients in the graphormer pipeline which are not part of the layer (e.g. the edge encodings) https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf

**Comment by CarloLucibello** [comment]
Anyways, this looks very good already

**Comment by 5hv5hvnk** [comment]
@CarloLucibello where should I add the tests ? Secondly is adding tests here similar too `PyTest` ?

**Comment by CarloLucibello** [comment]
take a look at the test folder. The entry point is `test/runtests.jl`. All julia packages are similarly structured.

**Comment by 5hv5hvnk** [comment]
Hi @CarloLucibello due to some conflicts that I was trying to resolve the PR is closed now. Let me try fix it and raise another PR

**Comment by CarloLucibello** [comment]
Sure, sorry for that, I made some pretty extensive changes

**Comment by 5hv5hvnk** [comment]
No problem you can have a look on #275

**Pull request closed without merging**